### PR TITLE
Short string representation of TypePointer in AST JSON

### LINKED ISSUE #3567: AST JSON: type name location is `storage` although it should be something like `unspecified`
The AST json of the contract
```
contract c {
    function f() {
        uint[] memory x;
    }
}
```
contains a part
```
{
{
    "constant" : false,
    "id" : 6,
    "name" : "x",
    "nodeType" : "VariableDeclaration",
    "scope" : 9,
    "src" : "28:15:0",
    "stateVariable" : false,
    "storageLocation" : "memory",
    "typeDescriptions" : 
    {
        "typeIdentifier" : "t_array$_t_uint256_$dyn_memory_ptr",
        "typeString" : "uint256[] memory"
    },
    "typeName" : 
    {
        "baseType" : 
        {
            "id" : 4,
            "name" : "uint",
            "nodeType" : "ElementaryTypeName",
            "src" : "28:4:0",
            "typeDescriptions" : 
            {
                "typeIdentifier" : "t_uint256",
                "typeString" : "uint256"
            }
        },
        "id" : 5,
        "length" : null,
        "nodeType" : "ArrayTypeName",
        "src" : "28:6:0",
        "typeDescriptions" : 
        {
            "typeIdentifier" : "t_array$_t_uint256_$dyn_storage_ptr",
            "typeString" : "uint256[] storage pointer"
        }
    },
    "value" : null,
    "visibility" : "internal"
}
```

So the type name has a type which has a `storage` location, while the variable itself has `memory` location. The reason behind this is that the location specifier is parsed as part of the variable declaration and not the type name (we do not want multiple different location specifiers as part of a type).

The solution to that could be to just omit the `storage pointer` part of the type by using the `_short` parameter of the `toString()` method in that case.

Discovered by @gnidan

### PR DESCRIPTION
**Pull request by erak**
Fixes #3567

### CODE REVIEW

**Review by axic — CHANGES_REQUESTED** [review]
Please adjusts (or add) tests for this.

**Comment by erak** [comment]
@axic Added missing tests.

**Review by chriseth — DISMISSED** [review]

**Comment by chriseth** [comment]
Looks good to me!

**Review by axic — CHANGES_REQUESTED** [review]
Please check the resulting `typeString` (and `typeIdentifier`). It is a good idea to take all examples from the initial issue.

**Comment by bit-shift** [comment]
- [ ] Add tests for ``typeString``
- [ ] Add tests for complex use case (see issue)

**Comment by erak** [comment]
@axic I've realised that the tests used the legacy AST output instead of the new compact one. That's the reason why checks for ``typeName`` and ``typeString`` were missing. I've changed that...

Not sure about ``t_array$_t_array$_t_uint256_$dyn_storage_$dyn_storage_ptr`` for ``uint[][] memory rows;``. It still looks wrong, but I wasn't sure how to fix the ``richIdentifier`` for ``ArrayType``.

@chriseth What do you think about this remaing issue?

**Comment by chriseth** [comment]
The rich identifier is fine, it is only important that it is really unique. They are not really exposed to the user as much as the other stuff is.

**Review by axic — DISMISSED** [review]
Please have at least two cases, one which triggers `_short = false` and one which triggers `_short = true`.

**Comment by axic** [comment]
Basically just extend all the other test cases with `typeIdentifier` expectation if they define a type.

**Comment by erak** [comment]
@axic The JSON AST tests are using the legacy JSON output. That's why you won't find ``typeIdentifier`` but rather ``typeName`` and ``type``. Actually those do not cover my changes since they only apply when using ``solc --ast-compact-json``. I think changing the tests such that they use the new output could be done while deprecating the legacy output for 0.5.0.

I think I'd rather add some tests that cover non-shortened type pointer as you suggested.

**Comment by axic** [comment]
> The JSON AST tests are using the legacy JSON output. 

In that case I think we should introduce two sub suites to group the legacy and the "compact" tests.

**Comment by erak** [comment]
Great idea, will do that!

**Comment by axic** [comment]
I actually meant that there can be multiple sub-levels of `BOOST_AUTO_TEST_SUITE(<name>)` within a single suite.

Also the case `documentation` contains both legacy and non-legacy tests, probably needs to be split into two.

**Comment by erak** [comment]
@axic Ok, my idea was to split them into 2 files right away, in order to not end up with a very large one at some point. But if you prefer a single file with multiple sub-suites, I'll put it back. I don't have a stong opinion on that.

I've pulled out the legacy part of the ``documentation`` test. Did I miss something?

**Comment by axic** [comment]
I don't have any preference, but if it is a split file I'd name it `ASTLegacyJSON`.

**Comment by erak** [comment]
@axic I've renamed the legacy file / suite and added two tests that cover long ``typeDescriptions`` for binary operations and identifiers.

**Comment by chriseth** [comment]
Please add a changelog entry.

**Review by ekpyron — DISMISSED** [review]
Looks fine to me now!

**Comment by ekpyron** [comment]
Rebased.

File: Changelog.md [hunk 1]
```diff
@@ -8,6 +8,7 @@ Features:
  * General: Limit the number of errors output in a single run to 256.
  * General: Support accessing dynamic return data in post-byzantium EVMs.
  * Interfaces: Allow overriding external functions in interfaces with public in an implementing contract.
+ * JSON AST: Use short string representation for TypePointers.
```
Comment (chriseth) [inline Changelog.md]: This should be a bugfix, and perhaps also phrased differently for the changelog: Remove storage qualifier for type name strings.

**Comment by chriseth** [comment]
Sorry, please rebase.

**Review by chriseth — APPROVED** [review]

**Pull request merged**